### PR TITLE
Update caching best practices for specific dependency managers

### PR DESCRIPTION
**Pull request by smart-alek**
Resolves [this JIRA issue](https://circleci.atlassian.net/browse/CIRCLE-11490).

This PR attempts to document how different dependency managers deal with partial cache restoration. I have almost certainly muddled a few things up in here, so I'm counting on both @eddiewebb and @michelle-luna to catch me where that's happened!.

### CODE REVIEW

File: jekyll/_cci2/caching.md [hunk 1]
```diff
@@ -236,46 +223,228 @@ The following example demonstrates how to use `restore_cache` and `save_cache` t
...
+
+To prevent this behavior,
+add a step
+that cleans Bundler
+before restoring dependencies from cache.
+
+{% raw %}
+
+```yaml
+steps:
```
[the remark this answers is not in the document]
Comment (smart-alek) [reply jekyll/_cci2/caching.md]: Which file? `Gemfile.lock`?

File: jekyll/_cci2/caching.md [hunk 2]
```diff
@@ -133,21 +133,7 @@ When a dependency tool runs,
 it would discover outdated dependencies and update them.
 This is referred to as a **partial cache restore**.
 
-### Clearing Cache
```
[the remark this answers is not in the document]
Comment (michelle-luna) [reply jekyll/_cci2/caching.md]: Having a high-level overview at the top with more detailed sections below is a good model to follow to help people get this concept and get all the keywords into headings, so again, please add back this section. The new sections are great, and the old section should stay there as well.

[the remark this answers is not in the document]
Comment (smart-alek) [reply jekyll/_cci2/caching.md]: I agree that a high-level overview at the top is a good model. In this case, though, **Clearing Cache** isn't high-level, at the top, or even an overview; it's in an entirely different section than the new section I added.

Will add the section back. From this feedback, my understanding is that it is acceptable to have information repeated (and in this case copied) in several places in a document.

[the remark this answers is not in the document]
Comment (michelle-luna) [reply jekyll/_cci2/caching.md]: Correct, thank you!

File: jekyll/_cci2/caching.md [hunk 3]
```diff
@@ -236,46 +237,310 @@ The following example demonstrates how to use `restore_cache` and `save_cache` t
...
+
+**Safe to Use Partial Cache Restoration?**
+Yes.
+
+Gradle repositories are intended
+to be centralized, shared, and massive.
+Partial caches can be restored
+without impacting which libraries
+are actually added to classpaths of generated artifacts.
+
```
Comment (michelle-luna) [inline jekyll/_cci2/caching.md]: this needs the raw/endraw tags added

File: jekyll/_cci2/caching.md [hunk 4]
```diff
@@ -236,46 +237,314 @@ The following example demonstrates how to use `restore_cache` and `save_cache` t
...
+  - save_cache:
+      paths:
+        - ~/.cache/yarn
+      key: yarn-packages-v1-{{ .Branch }}-{{ checksum "yarn.lock" }}
+```
+
+{% endraw %}
 
-Listing multiple keys for restoring a cache increases the odds of a partial cache hit. However, broadening your `restore_cache` scope to a wider history increases the risk of confusing failures. For example, if you have dependencies for Node v6 on an upgrade branch, but your other branches are still on Node v5, a `restore_cache` step that searches other branches might restore incompatible dependencies.
+## Caching Strategy Tradeoffs
```
Comment (smart-alek) [inline jekyll/_cci2/caching.md]: This isn't new — just got carried away with semantic linefeeds because I thought I'd be using copy from this section. 😬

**Pull request merged**